Adding Vulkan-style statistics counters where available

gary-sweet opened the pull request:
## Description

Previously, the 'stats' interface only used HWCPipe to interrogate performance counters from the device. HWCPipe is not widely supported, so this really only worked for Arm devices.

This change allows multiple 'StatsProviders' to be used. Each provider can supply all, some, or none of the requested counters depending on its capabilities. Any counters not supplied by one provider will be requested of the next in the list until one is found, or we run out of providers.

A second provider that uses the Vulkan performance counter extension has  been added and will be used where supported if HWCPipe cannot provide an equivalent counter. CPU counters can come from HWCPipe for example, whilst GPU counters come from the Vulkan extension. If TimestampQueries are available, they are used in the Vulkan-extension provider to more accurately measure the time spent executing the command buffer, which should match the time spent collecting counters.

The StatsProvider defines an abstract interface that is implemented by the concrete providers. The parameters used for graphing the statistics have moved from the GUI classes into the StatsProvider also. This allows providers to modify the names, units and scaling of counters where these differ
from the default values. It also feels more logical that the stats themselves should know best how to present themselves.

Since the Vulkan extension uses query pools which have to be begun and ended inside CommandBuffers there have been a few small changes to the examples to allow that to happen.

I have only tested this on embedded Linux.

Fixes #84 

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

gary-sweet committed:
Adding Vulkan-style statistics counters where available

Previously, the 'stats' interface just used HWCPipe to
interrogate performance counters from the device. HWCPipe
is not widely supported, so this really only worked for
Arm devices.

This change allows multiple 'StatsProviders' to be used.
Each provider can supply all, some, or none of the requested
counters depending on its capabilities. Any counters not
supplied by one provider will be requested of the next in
the list until one is found, or we run out of providers.
So, CPU counters can come from HWCPipe for example, whilst
GPU counters come from Vulkan's extension which has been
added as another provider.

The StatsProvider defines an abstract interface that is
implemented by the concrete providers. The parameters used
for graphing the statistics have moved from the GUI classes
into the StatsProvider also. This allows providers to modify
the names, units and scaling of counters where these differ
from the default values.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=120) <br/>All committers have signed the CLA.

gary-sweet commented:
I'd like to sign the CLA but https://cla-assistant.io/ is not responding.

gary-sweet committed:
Fixed up doxygen and copyright warnings

gary-sweet committed:
Hopefully fixed the Windows build

ZandroFargnoli commented on framework/core/device.cpp:
```diff
@@ -68,13 +68,48 @@ Device::Device(const PhysicalDevice &gpu, VkSurfaceKHR surface, std::unordered_m
+		supported_features.pNext = &perf_features;
+
+		// Check the feature support bits
+		vkGetPhysicalDeviceFeatures2KHR(gpu.get_handle(), &supported_features);
```
The PhysicalDevice class should be able to handle this so that here we can request the features

gary-sweet replied:
Ok - I'll take a look

gary-sweet replied:
Done

ZandroFargnoli commented on framework/core/device.cpp:
```diff
@@ -123,16 +158,34 @@ Device::Device(const PhysicalDevice &gpu, VkSurfaceKHR surface, std::unordered_m
 
+	const auto requested_gpu_features = gpu.get_requested_features();
+
+	if (has_performance_query)
```
If we let the GPU construct the linked list for us, it removes the need for us to check flags in this way, and we can just do `create_info.pNext  = gpu.get_requested_extension_features();`

gary-sweet replied:
Done

ZandroFargnoli commented on framework/gui.h:
```diff
@@ -187,31 +187,10 @@ class Gui
 		/**
-		 * @brief Per-statistic graph data
+		 * @brief Constructs a StatsView
+    	 * @param stats Const pointer to the Stats data object; may be null
```
Format

gary-sweet replied:
Oh, it looks like clang-format doesn't do anything with these comment blocks.
I originally thought it had just got it wrong. I'll fix them up.

ZandroFargnoli commented on framework/stats/frame_time_stats_provider.h:
```diff
@@ -0,0 +1,46 @@
+	 * @param delta_time Time since last sample
+	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	Sample sample(float delta_time, uint32_t active_frame_idx) override;
```
The word sample is confusing with the `VulkanSample`, it sounds like a base class for this, could it be renamed?

gary-sweet replied:
I'll try to think of something

gary-sweet replied:
I went with 'Counters'. Done

ZandroFargnoli commented on framework/stats/hwcpipe_stats_provider.h:
```diff
@@ -0,0 +1,130 @@
+
+		StatData() = default;
+
+		/**
```
Format

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/stats.cpp:
```diff
@@ -0,0 +1,273 @@
+#include "frame_time_stats_provider.h"
+#include "hwcpipe_stats_provider.h"
+#include "vulkan_stats_provider.h"
+namespace vkb
```
Newline?

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/stats.cpp:
```diff
@@ -0,0 +1,273 @@
+				}
+			}
+
+			if (pending_samples.size() == 0)
```
Throughout the project we use curly braces in our conditionals - even if they are one liners. I understand that by the guidelines we use that this is legal, however it's making this code seem inconsistent with the rest of the project. Seeing as the original code for these files had the braces could we keep it that way?

ZandroFargnoli replied:
Nitpick

gary-sweet replied:
I can go through and change all these if that's really necessary.
As you say, the guidelines don't specify this; I checked before I did it.
This is also done in several other places in the samples.

If you really want me to change all these, can I first ask that this is added to the coding style rules and the other places get fixed up too.

ZandroFargnoli replied:
I think it's okay, but there are a couple instances you breach the style guidelines but I've commented them, this can be resolved

ZandroFargnoli commented on framework/stats/stats.h:
```diff
@@ -0,0 +1,172 @@
+	 * @param cb The command buffer
+ 	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	void command_buffer_begun(CommandBuffer &cb, uint32_t active_frame_idx);
```
This function essentially prepares our providers to start collecting stats, I think this function name should be renamed to `begin` and the later `end`.

ZandroFargnoli replied:
The @brief should be changed respectively to detail this more accurately, along with when it should be called

JoseEmilio-ARM replied:
It's a tricky one, because for HWCPipe these are no-ops if I understood correctly (and they should be, we are reading system-wide counters), but for performance query we do need to collect data for a given command buffer

gary-sweet replied:
Exactly. They don't begin or end anything in hwcpipe so calling the abstract interface begin and end doesn't feel right to me.

ZandroFargnoli commented on framework/stats/stats_provider.cpp:
```diff
@@ -0,0 +1,53 @@
+namespace vkb
+{
+// Default graphing values for stats. May be overridden by individual providers.
+std::map<StatIndex, StatGraphData> StatsProvider::def_graph_map{
```
Nitpick but could this be renamed `default_graph_map`?

gary-sweet replied:
Yep - that was a bit lazy on my part :-)

ZandroFargnoli commented on framework/stats/vulkan_stats_provider.cpp:
```diff
@@ -0,0 +1,460 @@
+	perf_create_info.pCounterIndices   = counter_indices.data();
+
+	uint32_t passes_needed;
+	vkGetPhysicalDeviceQueueFamilyPerformanceQueryPassesKHR(gpu.get_handle(), &perf_create_info,
```
Could this function be part of the physical device class?

gary-sweet replied:
Done

ZandroFargnoli commented on framework/vulkan_sample.cpp:
```diff
@@ -190,16 +190,19 @@ void VulkanSample::update(float delta_time)
 
 	auto &command_buffer = render_context->begin();
 
+	uint32_t active_frame_idx = render_context->get_active_frame_index();
```
Perhaps add a comment here explaining that this section collects the performance data for the sample graphs, from todays meeting I've been reminded how vital this part is to the whole project and I think having some extra stress here would be nice

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/vulkan_stats_provider.h:
```diff
@@ -0,0 +1,163 @@
+	// The timestamp period
+	float timestamp_period{1.0f};
+
+	// And one for timestamps
```
Comment needs to be descriptive on its own

gary-sweet replied:
Done

ZandroFargnoli commented on framework/core/command_buffer.h:
```diff
@@ -219,6 +220,14 @@ class CommandBuffer
+
+	void end_query(const QueryPool &query_pool, uint32_t query);
+
+	void write_timestamp(VkPipelineStageFlagBits pipelineStage, const QueryPool &query_pool, uint32_t query);
```
snake_case

gary-sweet replied:
It's hard switching coding styles sometimes ;-)
Done

ZandroFargnoli commented on framework/core/command_buffer.cpp:
```diff
@@ -755,6 +755,27 @@ const bool CommandBuffer::is_render_size_optimal(const VkExtent2D &framebuffer_e
+	vkCmdEndQuery(get_handle(), query_pool.get_handle(), query);
+}
+
+void CommandBuffer::write_timestamp(VkPipelineStageFlagBits pipelineStage,
```
snake_case

gary-sweet replied:
Done

ZandroFargnoli commented on framework/core/query_pool.cpp:
```diff
@@ -0,0 +1,67 @@
+	return handle;
+}
+
+void QueryPool::host_reset(uint32_t firstQuery, uint32_t queryCount)
```
snake_case

gary-sweet replied:
Done

ZandroFargnoli commented on framework/core/query_pool.h:
```diff
@@ -0,0 +1,80 @@
+	 * @param firstQuery The first query to reset
+	 * @param queryCount The number of queries to reset
+	 */
+	void host_reset(uint32_t firstQuery, uint32_t queryCount);
```
snake_case

gary-sweet replied:
Done

ZandroFargnoli reviewed (commented):
Really good work Gary, I've left some comments around the formatting and code style of the MR. I've tested the code base and it works on Windows and Android (S9/S10/S20). Happy to approve once the comments have been resolved.

JoseEmilio-ARM commented on framework/gui.h:
```diff
@@ -225,83 +204,7 @@ class Gui
 		void reset_max_value(const StatIndex index);
 
 		/// Per-statistic max values
```
We could remove this comment, I think it was already outdated :)

gary-sweet replied:
Done

JoseEmilio-ARM commented on framework/stats/frame_time_stats_provider.cpp:
```diff
@@ -0,0 +1,45 @@
+
+#include "frame_time_stats_provider.h"
+
+#include <cassert>
```
I think we generally use `throw VulkanException` instead

gary-sweet replied:
This include wasn't needed anyway

JoseEmilio-ARM commented on framework/stats/hwcpipe_stats_provider.cpp:
```diff
@@ -0,0 +1,218 @@
+	assert(is_available(index) && "HWCPipeStatsProvider::get_graph_data() called with invalid StatIndex");
+
+	// HWCPipe reports combined vertex/compute cycles (which is Arm specific)
+	// Ensure we report graph that with the correct name when asked for vertex cycles
```
extra 'that' in the comment

Also maybe move the assert (or VulkanException :) ) up to be the first line in the function?

gary-sweet replied:
From what I can see, VulkanException is generally thrown when a Vulkan call or Vulkan creation fails. This assert is really checking that the framework isn't calling this without first checking for its availability. It doesn't feel in the same class as a VulkanException to me. Since it should never happen an assert, which only happens in debug builds, feels more correct.

gary-sweet replied:
I've moved the assert though.

JoseEmilio-ARM replied:
Yes that makes sense, thanks!

JoseEmilio-ARM commented on framework/stats/hwcpipe_stats_provider.h:
```diff
@@ -0,0 +1,130 @@
+	/**
+	 * @brief Retrieve a new sample set from polled sampling
+	 * @param delta_time Time since last sample
+	 * @param active_frame_idx Which of the framebuffers is active
```
Could we document here that HWCPipe does not use `active_frame_idx`?

gary-sweet replied:
Done

JoseEmilio-ARM commented on framework/stats/stats_common.h:
```diff
@@ -0,0 +1,120 @@
+{
+  public:
+	/**
+		 * @brief Constructs data for the graph
```
extra indentation

gary-sweet replied:
I expected clang-format to take care of that. It seems it doesn't though.
Done

JoseEmilio-ARM commented on framework/stats/vulkan_stats_provider.cpp:
```diff
@@ -0,0 +1,460 @@
+		// NOTE: The names here are actually regular-expressions.
+		// Counter names can change between hardware variants for the same vendor,
+		// so regular expression names mean that multiple h/w variants can be easily supported.
+		// clang-format off
```
why do we turn clang-format off in a few places? how does the output look like, and can we fix it by adjusting the .clang-format file?

gary-sweet replied:
I've turned off clang-formatting for a few places where tables are being defined in the code.
Without doing so the table is just tightly aligned which makes it much harder to read in my opinion:
```
		vendor_data = {
		    {StatIndex::gpu_cycles, {"cycle_count"}},
		    {StatIndex::gpu_vertex_cycles, {"qpu_active_cycles_vertex_coord_user"}},
		    {StatIndex::gpu_fragment_cycles, {"qpu_active_cycles_fragment"}},
		    {StatIndex::gpu_fragment_jobs, {"render_jobs_completed"}},
		    {StatIndex::gpu_ext_reads, {"gpu_mem_reads"}},
		    {StatIndex::gpu_ext_writes, {"gpu_mem_writes"}},
		    {StatIndex::gpu_ext_read_bytes, {"gpu_bytes_read"}},
		    {StatIndex::gpu_ext_write_bytes, {"gpu_bytes_written"}},
		};
```
compared with:
```
		vendor_data = {
		    {StatIndex::gpu_cycles,          {"cycle_count"}},
		    {StatIndex::gpu_vertex_cycles,   {"qpu_active_cycles_vertex_coord_user"}},
		    {StatIndex::gpu_fragment_cycles, {"qpu_active_cycles_fragment"}},
		    {StatIndex::gpu_fragment_jobs,   {"render_jobs_completed"}},
		    {StatIndex::gpu_ext_reads,       {"gpu_mem_reads"}},
		    {StatIndex::gpu_ext_writes,      {"gpu_mem_writes"}},
		    {StatIndex::gpu_ext_read_bytes,  {"gpu_bytes_read"}},
		    {StatIndex::gpu_ext_write_bytes, {"gpu_bytes_written"}},
		};
```
There is no way to do this in the clang-format file as far as I'm aware.

JoseEmilio-ARM replied:
Yes I agree, thanks!

SaschaWillems commented:
This looks good to me. I just ran this on my desktop with an RTX 2060 and all the samples I checked worked as expected. Can't check Vulkan perf counters though as the extension is not available.

gary-sweet committed:
Addressed some formatting and naming feedback

gary-sweet committed:
Moved some things into physical device

Simplifies code in device

JoseEmilio-ARM commented on framework/core/device.cpp:
```diff
@@ -253,7 +227,7 @@ Device::Device(const PhysicalDevice &gpu, VkSurfaceKHR surface, std::unordered_m
 	command_pool = std::make_unique<CommandPool>(*this, get_queue_by_flags(VK_QUEUE_GRAPHICS_BIT | VK_QUEUE_COMPUTE_BIT, 0).get_family_index());
 	fence_pool   = std::make_unique<FencePool>(*this);
-}
+}        // namespace vkb
```
was this accidentally added by clang-format?

gary-sweet replied:
If I remove it and clang-format the file it comes back. So deliberate by clang-format I guess.

JoseEmilio-ARM replied:
yes unfortunately we had this issue before, annoying because that bracket doesn't close the namespace 😞

gary-sweet replied:
Are you sure? It looks like the namespace closure to me

JoseEmilio-ARM replied:
that's the one in line 622 that was there before, but this one is in line 230

gary-sweet replied:
Ah yes - hadn't noticed the line number wasn't the end of the file.
That one does go away, so clang-format must have added it when I had some mismatched braces during development I guess.

I'll push a fix. Thanks

gary-sweet replied:
Done

gary-sweet committed:
Fixed incorrect namespace comment

gary-sweet committed:
Tweak to Broadcom specific counters

gary-sweet committed:
Reworked top level stats interface and construction

Based on private feedback from Zandro, I've moved the
construction of the stats object into the base
VulkanSamples class which leaves individual samples
to just request the samples they would like.

The active_frame_index is no longer passed to
update_stats() or the command_buffer_begun/ended()
methods. It is fetched internally by the providers
that require it.

gary-sweet committed:
Fix Windows build

size_t -> uint32_t

ZandroFargnoli commented on framework/stats/hwcpipe_stats_provider.cpp:
```diff
@@ -0,0 +1,218 @@
+{
+HWCPipeStatsProvider::HWCPipeStatsProvider(std::set<StatIndex> & requested_stats,
+                                           CounterSamplingConfig sampling_config) :
+    sampling_config(sampling_config)
```
This appears to be unused in `HWCPipeStatsProvider`?

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/stats.cpp:
```diff
@@ -0,0 +1,287 @@
+	if (providers.size() != 0)
+	{
+		LOGE("Stats must only be requested once");
+		throw std::runtime_error("Stats must only be requested once");
```
A runtime error should be logged in `LOGE` already so we shouldn't need to write the same error message twice here

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/hwcpipe_stats_provider.cpp:
```diff
@@ -0,0 +1,218 @@
+					else
+						++iter;
+				}
+				else
```
If `if` have curly braces then `else` need to continue this as per [Google C++ Style Guide: Conditionals](https://google.github.io/styleguide/cppguide.html#Conditionals)

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/hwcpipe_stats_provider.cpp:
```diff
@@ -0,0 +1,218 @@
+					else
+						++iter;
+				}
+				else
```
Same here

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/stats.cpp:
```diff
@@ -0,0 +1,287 @@
+	}
+}
+
+void Stats::command_buffer_begun(CommandBuffer &cb)
```
Can these functions have a word like `begin_sampling` or `start`, I still dislike the name as these functions have to be placed all around the project. The description of the function should explain that the command buffer has to be in the recording state (i.e. begun), but the name of the function should only pertain to what the function is doing and not when it should be called

gary-sweet replied:
None of the names are strictly correct, but I've gone with begin/end_sampling

ZandroFargnoli commented on framework/stats/stats.cpp:
```diff
@@ -0,0 +1,287 @@
+		p->command_buffer_begun(cb);
+}
+
+void Stats::command_buffer_ending(CommandBuffer &cb)
```
Same goes for this, can this be `end_sampling`, or `finish` or something along these lines.

gary-sweet replied:
Done

ZandroFargnoli commented on framework/stats/vulkan_stats_provider.h:
```diff
@@ -0,0 +1,163 @@
+	 * @brief A command buffer that we want stats about has just begun
+	 * @param cb The command buffer
+	 */
+	void command_buffer_begun(CommandBuffer &cb) override;
```
I think that this should also be renamed in line with `Stats`

gary-sweet replied:
It has to be, it's a derived method. Done

ZandroFargnoli commented on framework/stats/vulkan_stats_provider.h:
```diff
@@ -0,0 +1,163 @@
+	 * @brief A command buffer that we want stats about is about to be ended
+	 * @param cb The command buffer
+	 */
+	void command_buffer_ending(CommandBuffer &cb) override;
```
And here

gary-sweet replied:
Done

ZandroFargnoli reviewed (changes requested):
Happy with the larger scale architectural changes you made that we discussed, a few smaller issues to resolve but otherwise LGTM

gary-sweet committed:
Tidy up some formatting and rename some methods

ZandroFargnoli reviewed (dismissed):
LGTM

AlejandroCosin reviewed (commented):
I tested all samples in a PowerVR GPU and all of them work fine but one: Wait Idle, which has a crash that happens for all platforms. I already submitted a pull request with the fix (just two lines of code https://github.com/gary-sweet/Vulkan-Samples/pull/2).

gary-sweet committed:
Fix for WaitIdle

Override prepare_render_context() to use custom render_context.
Ensures stats always sees the right render context.

gary-sweet commented:
@AlejandroCosin I've made an alternative fix for the WaitIdle problem you saw. Can you check that it has the same effect please. The original code works for me for some reason.

AlejandroCosin commented:
> @AlejandroCosin I've made an alternative fix for the WaitIdle problem you saw. Can you check that it has the same effect please. The original code works for me for some reason.

Sure! will test and tell you :)

AlejandroCosin commented:
@gary-sweet I tested in Windows and Android the Wait Idle Sample with your last changes ("Fix for WaitIdle" commit in your branch perf_counters) and everything works great
